src/Agent.Worker/Handlers/StepHost.cs

```diff
@@ -0,0 +1,126 @@
+                                            IDictionary<string, string> environment,
+                                            bool requireExitCodeZero,
+                                            Encoding outputEncoding,
+                                            bool killProcessOnCancel,
+                                            CancellationToken cancellationToken)
+        {
+            using (var processInvoker = HostContext.CreateService<IProcessInvoker>())
+            {
+                processInvoker.OutputDataReceived += OutputDataReceived;
+                processInvoker.ErrorDataReceived += ErrorDataReceived;
+
+                return await processInvoker.ExecuteAsync(workingDirectory: workingDirectory,
+                                                         fileName: fileName,
+                                                         arguments: arguments,
+                                                         environment: environment,
+                                                         requireExitCodeZero: requireExitCodeZero,
+                                                         outputEncoding: outputEncoding,
+                                                         killProcessOnCancel: killProcessOnCancel,
+                                                         cancellationToken: cancellationToken);
+            }
+        }
+    }
+
+    public sealed class ContainerStepHost : AgentService, IContainerStepHost
+    {
+        public ContainerInfo Container { get; set; }
+        public event EventHandler<ProcessDataReceivedEventArgs> OutputDataReceived;
+        public event EventHandler<ProcessDataReceivedEventArgs> ErrorDataReceived;
+
+        public async Task<int> ExecuteAsync(string workingDirectory,
+                                            string fileName,
+                                            string arguments,
+                                            IDictionary<string, string> environment,
+                                            bool requireExitCodeZero,
+                                            Encoding outputEncoding,
+                                            bool killProcessOnCancel,
+                                            CancellationToken cancellationToken)
+        {
+            // make sure container exist.
+            ArgUtil.NotNull(Container, nameof(Container));
+            ArgUtil.NotNullOrEmpty(Container.ContainerId, nameof(Container.ContainerId));
+
+            var dockerManger = HostContext.GetService<IDockerCommandManager>();
+            string containerEnginePath = dockerManger.DockerPath;
+
+            string envOptions = "";
+            foreach (var env in environment)
+            {
+                envOptions += $" -e \"{env.Key}={env.Value.Replace("\"", "\\\"")}\"";
+            }
+
+#if !OS_WINDOWS
+            string workingDirectoryEscaped = StringUtil.Format(@"""{0}""", workingDirectory);
```

Comment:
for linux replace `"` with `\"` for workingdir and filepath

Reply:
do we really need to escape? the handler is already adding `"` and replace `"` with `\"`